### PR TITLE
Handle incoming AuthnRequests

### PR DESCRIPTION
**Pull request by jkakavas**
- Expose an API that consumes (possibly signed) AuthnRequests as
defined by the HTTP-Redirect binding.
- Process AuthnRequests, validate and parse them into a minimum
set of information to be used for subsequent API calls to get a
SAML Response

### CODE REVIEW

File: x-pack/plugin/identity-provider/src/main/java/org/elasticsearch/xpack/idp/action/SamlValidateAuthnRequestRequest.java
Comment (tvernum) [inline x-pack/plugin/identity-provider/src/main/java/org/elasticsearch/xpack/idp/action/SamlValidateAuthnRequestRequest.java]: See similar comment on other PR.
I think there are two options that make sense:  
- either assume that there is only a single validation and simplify this down to `if (isNullOrEmpty) { return addValidationError(); } else { return null; }`  
- or follow the standard pattern we use here, in preparation for additional validations, and pass `validationException` in the 2nd argument.

File: x-pack/plugin/identity-provider/src/main/java/org/elasticsearch/xpack/idp/action/SamlValidateAuthnRequestResponse.java
Comment (tvernum) [inline x-pack/plugin/identity-provider/src/main/java/org/elasticsearch/xpack/idp/action/SamlValidateAuthnRequestResponse.java]: Can we have `null` checks here, since the serialization assumes that they're not null?

File: x-pack/plugin/identity-provider/src/main/java/org/elasticsearch/xpack/idp/action/SamlValidateAuthnRequestResponse.java
Comment (tvernum) [inline x-pack/plugin/identity-provider/src/main/java/org/elasticsearch/xpack/idp/action/SamlValidateAuthnRequestResponse.java]: These can be final.

File: x-pack/plugin/identity-provider/src/main/java/org/elasticsearch/xpack/idp/rest/RestSamlValidateAuthenticationRequestAction.java
Comment (tvernum) [inline x-pack/plugin/identity-provider/src/main/java/org/elasticsearch/xpack/idp/rest/RestSamlValidateAuthenticationRequestAction.java]: Per other PR, I think an object `{ "service_provider": { "entity_id": "..." } }` sets us up for better future expansion.

File: x-pack/plugin/identity-provider/src/main/java/org/elasticsearch/xpack/idp/rest/RestSamlValidateAuthenticationRequestAction.java
Comment (tvernum) [inline x-pack/plugin/identity-provider/src/main/java/org/elasticsearch/xpack/idp/rest/RestSamlValidateAuthenticationRequestAction.java]: I don't love this name. I wonder if we should be more explicit and call it `xxx_state` or something (`authn_state`, `request_state` ?)

File: x-pack/plugin/identity-provider/src/main/java/org/elasticsearch/xpack/idp/saml/authn/SamlAuthnRequestValidator.java
Comment (tvernum) [inline x-pack/plugin/identity-provider/src/main/java/org/elasticsearch/xpack/idp/saml/authn/SamlAuthnRequestValidator.java]: Is this needed? Can't we just call `decodeQueryString` and check `parameters.size()`   ?

File: x-pack/plugin/identity-provider/src/main/java/org/elasticsearch/xpack/idp/saml/authn/SamlAuthnRequestValidator.java
Comment (tvernum) [inline x-pack/plugin/identity-provider/src/main/java/org/elasticsearch/xpack/idp/saml/authn/SamlAuthnRequestValidator.java]: I don't understand why we have a lambda here. Can't we just pass a `ParameterizedMessage` to logAndThrow? We know the message will be instantiated since it's going to be thrown.

File: x-pack/plugin/identity-provider/src/main/java/org/elasticsearch/xpack/idp/saml/authn/SamlAuthnRequestValidator.java
Comment (tvernum) [inline x-pack/plugin/identity-provider/src/main/java/org/elasticsearch/xpack/idp/saml/authn/SamlAuthnRequestValidator.java]: It's a preference, but I think I'd find this neater to have a `SamlUtils.elementNameMatches(element, namespace, local)` method.

File: x-pack/plugin/identity-provider/src/main/java/org/elasticsearch/xpack/idp/saml/authn/SamlAuthnRequestValidator.java
Comment (tvernum) [inline x-pack/plugin/identity-provider/src/main/java/org/elasticsearch/xpack/idp/saml/authn/SamlAuthnRequestValidator.java]: The XML could be long, should we truncate at some reasonable point? We mostly only care about the namespace/name anyway.

File: x-pack/plugin/identity-provider/src/main/java/org/elasticsearch/xpack/idp/saml/authn/SamlAuthnRequestValidator.java
Comment (tvernum) [inline x-pack/plugin/identity-provider/src/main/java/org/elasticsearch/xpack/idp/saml/authn/SamlAuthnRequestValidator.java]: Per other PR, I think `getRegisteredServiceProvider(String)` will serve us better.

File: x-pack/plugin/identity-provider/src/main/java/org/elasticsearch/xpack/idp/saml/authn/SamlAuthnRequestValidator.java
Comment (tvernum) [inline x-pack/plugin/identity-provider/src/main/java/org/elasticsearch/xpack/idp/saml/authn/SamlAuthnRequestValidator.java]: I'm pretty sure this will end up async (needing to lookup SP in index).
Can we structure it to have a listener now, so that it's a bit simpler to change later on?

File: x-pack/plugin/identity-provider/src/main/java/org/elasticsearch/xpack/idp/saml/authn/SamlAuthnRequestValidator.java
Comment (tvernum) [inline x-pack/plugin/identity-provider/src/main/java/org/elasticsearch/xpack/idp/saml/authn/SamlAuthnRequestValidator.java]: This method in particular will probably need to end up async.

File: x-pack/plugin/identity-provider/src/main/java/org/elasticsearch/xpack/idp/saml/authn/SamlAuthnRequestValidator.java
Comment (tvernum) [inline x-pack/plugin/identity-provider/src/main/java/org/elasticsearch/xpack/idp/saml/authn/SamlAuthnRequestValidator.java]: There's nothing here for empty string, and I think it's probably worth having an explicit error for that case.
I don't fancy debugging the type of error messages we'd get from trying to process empty signatures/algs.

File: x-pack/plugin/identity-provider/src/main/java/org/elasticsearch/xpack/idp/saml/authn/SamlAuthnRequestValidator.java
Comment (tvernum) [inline x-pack/plugin/identity-provider/src/main/java/org/elasticsearch/xpack/idp/saml/authn/SamlAuthnRequestValidator.java]: I can check the spec if you don't know, but should `sigAlg != null && signature == null` be an error too?

File: x-pack/plugin/identity-provider/src/main/java/org/elasticsearch/xpack/idp/saml/authn/SamlAuthnRequestValidator.java
Comment (tvernum) [inline x-pack/plugin/identity-provider/src/main/java/org/elasticsearch/xpack/idp/saml/authn/SamlAuthnRequestValidator.java]: So we validate signatures, but we never care if a request is unsigned?
I don't know what we want (should an SP have a config for what sorts of messages they will sign, and we enforce signatures there?), but perhaps we at least need a `TODO` here.

File: x-pack/plugin/identity-provider/src/main/java/org/elasticsearch/xpack/idp/saml/authn/SamlAuthnRequestValidator.java
Comment (tvernum) [inline x-pack/plugin/identity-provider/src/main/java/org/elasticsearch/xpack/idp/saml/authn/SamlAuthnRequestValidator.java]: Are you intentionally saying that we would go through the user auth phase and _then_ return an error to the SP?
Is that necessary?

I think we should indicate to the caller that validation failed, and either:
1. return all the info they need to send back to the SP (but that seems like it duplicates the other API), _OR_
2. provide a new API that takes an error object from this response and turns it an Authn `Response` failure, _OR_
3. make it possible to call the `Response` API without user authc, if-and-only-if the state data indicates an error.

File: x-pack/plugin/identity-provider/src/main/java/org/elasticsearch/xpack/idp/saml/authn/SamlAuthnRequestValidator.java
Comment (tvernum) [inline x-pack/plugin/identity-provider/src/main/java/org/elasticsearch/xpack/idp/saml/authn/SamlAuthnRequestValidator.java]: Shouldn't this reference `sigAlg`? or at least, should we validate & require that `sigAlg` is SHA256+RSA? (Ir we only want to support 1 algorithm).

**Review by tvernum — COMMENTED** [review]
I didn't finish the review, but I think there's enough to go on for now.

File: x-pack/plugin/identity-provider/src/main/java/org/elasticsearch/xpack/idp/rest/RestSamlValidateAuthenticationRequestAction.java
Comment (jkakavas) [reply x-pack/plugin/identity-provider/src/main/java/org/elasticsearch/xpack/idp/rest/RestSamlValidateAuthenticationRequestAction.java]: >I don't love this name.

That makes two of us. `xxx_state` works fine for me

File: x-pack/plugin/identity-provider/src/main/java/org/elasticsearch/xpack/idp/saml/authn/SamlAuthnRequestValidator.java
Comment (jkakavas) [reply x-pack/plugin/identity-provider/src/main/java/org/elasticsearch/xpack/idp/saml/authn/SamlAuthnRequestValidator.java]: leftover from before I discovered `RestUtils.decodeQueryString(queryString, 0, parameters);`

File: x-pack/plugin/identity-provider/src/main/java/org/elasticsearch/xpack/idp/saml/authn/SamlAuthnRequestValidator.java
Comment (jkakavas) [reply x-pack/plugin/identity-provider/src/main/java/org/elasticsearch/xpack/idp/saml/authn/SamlAuthnRequestValidator.java]: True, I carried this over from previous version where I wouldn't log_and_throw and it made sense then

File: x-pack/plugin/identity-provider/src/main/java/org/elasticsearch/xpack/idp/saml/authn/SamlAuthnRequestValidator.java
Comment (jkakavas) [reply x-pack/plugin/identity-provider/src/main/java/org/elasticsearch/xpack/idp/saml/authn/SamlAuthnRequestValidator.java]: I left this on purpose for the registration API PR but no objections to tackle this here, will do!

File: x-pack/plugin/identity-provider/src/main/java/org/elasticsearch/xpack/idp/saml/authn/SamlAuthnRequestValidator.java
Comment (jkakavas) [reply x-pack/plugin/identity-provider/src/main/java/org/elasticsearch/xpack/idp/saml/authn/SamlAuthnRequestValidator.java]: Correct, it's either both or none

File: x-pack/plugin/identity-provider/src/main/java/org/elasticsearch/xpack/idp/saml/authn/SamlAuthnRequestValidator.java
Comment (jkakavas) [reply x-pack/plugin/identity-provider/src/main/java/org/elasticsearch/xpack/idp/saml/authn/SamlAuthnRequestValidator.java]: Good catch, I can already imagine the debugging session

File: x-pack/plugin/identity-provider/src/main/java/org/elasticsearch/xpack/idp/saml/authn/SamlAuthnRequestValidator.java
Comment (jkakavas) [reply x-pack/plugin/identity-provider/src/main/java/org/elasticsearch/xpack/idp/saml/authn/SamlAuthnRequestValidator.java]: This can go a few ways: 
- We can demand _only_ signed AuthnRequests, declare that in our metadata too `WantAuthnRequestsSigned` and fail here. 
- Allow SPs to sign and handle this but not mandate it ( what is what we have so far )
- Set this in configuration when an SP is registered and act accordingly. 

I'll add a TODO so that we don't miss this and we can discuss

File: x-pack/plugin/identity-provider/src/main/java/org/elasticsearch/xpack/idp/saml/authn/SamlAuthnRequestValidator.java
Comment (jkakavas) [reply x-pack/plugin/identity-provider/src/main/java/org/elasticsearch/xpack/idp/saml/authn/SamlAuthnRequestValidator.java]: > Are you intentionally saying that we would go through the user auth phase and then return an error to the SP?
Is that necessary?

Yes. If the user can't authenticate to the IDP we should not return `urn:oasis:names:tc:SAML:2.0:status:InvalidNameIdPolicy` but `urn:oasis:names:tc:SAML:2.0:status:AuthnFailed` . Specific statues are optional in the spec and we can select to ignore them but it's very useful for debugging and important for interoperability. 

> return all the info they need to send back to the SP (but that seems like it duplicates the other API)

That would require the caller of the API to be able to construct SAML Responses and I don't think we should expect that. 

Between 2 and 3 I was leaning towards 3 and that we would handle that in the Transport action of the other API, so that we don't need to build additional logic in the UI/caller of the API. I'm ofcourse open to discuss this further

File: x-pack/plugin/identity-provider/src/main/java/org/elasticsearch/xpack/idp/saml/authn/SamlAuthnRequestValidator.java
Comment (jkakavas) [reply x-pack/plugin/identity-provider/src/main/java/org/elasticsearch/xpack/idp/saml/authn/SamlAuthnRequestValidator.java]: :man_facepalming:

File: x-pack/plugin/identity-provider/src/main/java/org/elasticsearch/xpack/idp/saml/authn/SamlAuthnRequestValidator.java
Comment (jkakavas) [reply x-pack/plugin/identity-provider/src/main/java/org/elasticsearch/xpack/idp/saml/authn/SamlAuthnRequestValidator.java]: Based on our discussion today, I went with the following: 

- If the SP should sign requests, this should be explicitly defined during registration
- If the SP should sign requests, and the request has a signature, we validate it
- If the SP should sign requests, and the request doesn't have a signature, we fail
- If the SP should not sign requests, we do not attempt to parse and validate a signature even if one exists

**Comment by jkakavas** [comment]
> address feedbacl and add Transport action tests

Mixed up commit messages thinking about https://github.com/elastic/elasticsearch/pull/51830. I didn't add Transport action tests here as all it does in `doExecute` is to call `processQueryString`

**Comment by tvernum** [comment]
I'm going to hold off re-reviewing this until the other handler PR is merged. I think we're good, but the merge is going to be a bit messy.

**Comment by jkakavas** [comment]
> I'm going to hold off re-reviewing this until the other handler PR is merged. I think we're good, but the merge is going to be a bit messy.

This is ready @tvernum

File: x-pack/plugin/identity-provider/src/main/java/org/elasticsearch/xpack/idp/action/SamlValidateAuthnRequestResponse.java
Comment (tvernum) [inline x-pack/plugin/identity-provider/src/main/java/org/elasticsearch/xpack/idp/action/SamlValidateAuthnRequestResponse.java]: I think we probably want a `Map.copyOf` here.
If the provided `Map` is already immutable, then that's a (more or less) zero cost operation.

File: x-pack/plugin/identity-provider/src/main/java/org/elasticsearch/xpack/idp/saml/authn/SamlAuthnRequestValidator.java
Comment (tvernum) [inline x-pack/plugin/identity-provider/src/main/java/org/elasticsearch/xpack/idp/saml/authn/SamlAuthnRequestValidator.java]: Can we add `(empty string)` or `(zero parameters)` or something like that to this message.
From the code it will be clear, but it won't be obvious to the person reading the logs.

File: x-pack/plugin/identity-provider/src/main/java/org/elasticsearch/xpack/idp/saml/authn/SamlAuthnRequestValidator.java
Comment (tvernum) [inline x-pack/plugin/identity-provider/src/main/java/org/elasticsearch/xpack/idp/saml/authn/SamlAuthnRequestValidator.java]:
```suggestion
                return;
            }
```

File: x-pack/plugin/identity-provider/src/main/java/org/elasticsearch/xpack/idp/saml/authn/SamlAuthnRequestValidator.java
Comment (tvernum) [inline x-pack/plugin/identity-provider/src/main/java/org/elasticsearch/xpack/idp/saml/authn/SamlAuthnRequestValidator.java]: Should we log something? `validateSignature` throws an exception without logging it (e.g. if there's an invalid signature algorithm), but we don't log it here either.

File: x-pack/plugin/identity-provider/src/main/java/org/elasticsearch/xpack/idp/saml/authn/SamlAuthnRequestValidator.java
Comment (tvernum) [inline x-pack/plugin/identity-provider/src/main/java/org/elasticsearch/xpack/idp/saml/authn/SamlAuthnRequestValidator.java]: Can we move this (`"redirect"`) to a constant?

File: x-pack/plugin/identity-provider/src/main/java/org/elasticsearch/xpack/idp/saml/authn/SuccessfulAuthenticationResponseMessageBuilder.java
Comment (tvernum) [inline x-pack/plugin/identity-provider/src/main/java/org/elasticsearch/xpack/idp/saml/authn/SuccessfulAuthenticationResponseMessageBuilder.java]: What's the reason for dropping the `SamlFactory` ?

File: x-pack/plugin/identity-provider/src/main/java/org/elasticsearch/xpack/idp/saml/authn/SamlAuthnRequestValidator.java
Comment (jkakavas) [reply x-pack/plugin/identity-provider/src/main/java/org/elasticsearch/xpack/idp/saml/authn/SamlAuthnRequestValidator.java]: Yes, I handled ESE here explicitly so that we don't wrap it in a new ESE in logAndRespond but forgot to log :/

File: x-pack/plugin/identity-provider/src/main/java/org/elasticsearch/xpack/idp/saml/authn/SamlAuthnRequestValidator.java
Comment (jkakavas) [reply x-pack/plugin/identity-provider/src/main/java/org/elasticsearch/xpack/idp/saml/authn/SamlAuthnRequestValidator.java]: Yes, I've already changed it in https://github.com/elastic/elasticsearch/pull/52415 to use `org.opensaml.saml.common.xml.SAMLConstants.SAML2_REDIRECT_BINDING_URI` instead of simple strings, will adjust it here too

File: x-pack/plugin/identity-provider/src/main/java/org/elasticsearch/xpack/idp/saml/authn/SuccessfulAuthenticationResponseMessageBuilder.java
Comment (jkakavas) [reply x-pack/plugin/identity-provider/src/main/java/org/elasticsearch/xpack/idp/saml/authn/SuccessfulAuthenticationResponseMessageBuilder.java]: I thought that SamlUtils covers the functionality and purpose so there was no reason to have both and I moved relevant parts from SamlFactory to SamlUtils ( this is more or less what we do in the saml realm implementation ). 

If you feel that we should split somehow ( Instance with initialization for loggers in Factory and all static methods in Utils for instance ) I'm happy to discuss or just do it. I have no strong preference

Comment (tvernum) [reply x-pack/plugin/identity-provider/src/main/java/org/elasticsearch/xpack/idp/saml/authn/SuccessfulAuthenticationResponseMessageBuilder.java]: I found SamlUtils in security to have been a bit overloaded, and was hoping to avoid replicating it here. I'm not stuck on it, I just have an aversion to catch-all utility classes.

Comment (jkakavas) [reply x-pack/plugin/identity-provider/src/main/java/org/elasticsearch/xpack/idp/saml/authn/SuccessfulAuthenticationResponseMessageBuilder.java]: Thinking this through again and seeing that I've made all methods in SamlUtils static already, I'm not sure what distinction would make sense. As in what should be in Utils and what should be in the Factory class. At 300 loc, SamlUtils doesn't seem too bloated atm.

I'm open to suggestions on what should be split away and back to a Factory class, or we could merge as is and spit these somehow once/if SamlUtils gets too bloated/overloaded

Comment (tvernum) [reply x-pack/plugin/identity-provider/src/main/java/org/elasticsearch/xpack/idp/saml/authn/SuccessfulAuthenticationResponseMessageBuilder.java]: This is definitely a personal preference thing, so I'm OK with merging as-is (though, in terms of preferences, I preferred the existing code).

Issues with Utility classes:
1. They become a catch-all dumping ground that becomes a cheap answer when something is needed, and avoids doing the harder design thinking. A good chunk of Utility class methods are working around broken class design that ideally would be fixed up front (but sometimes you have to because the broken class is in the JDK).
2. Related to the above, they often end up with a weird mix of responsibilities/behaviours. Because they're just chunks of code that _"do something, that someone thought was useful"_, they typically don't go through the process of _Is this the right behaviour for this method?_. 
3. Related to that, they can be hard to test well, unless they've got a clear interface & a single purpose (those are the best kind of utility method).
4. More importantly, it becomes hard to test code the _uses_ utility methods. They can't be mocked (because they're static), and it's often hard to trigger some of the behaviours that you need. As an example, while writing tests for SSL error messages, it was almost impossible to force PemUtils to throw certain exceptions so that I could verify they were handled neatly.
5. It can be hard to find the right method, because of the dumping ground problem. Ever tried to find the right helper method in amongst `XContentUtils`, `XContentParserUtils`, `XContentHelper`, `XContentTestUtils` ?

Now, you're probably thinking _Wait, `SamlUtils` and `SamlFactory` makes problem 5 **worse**_. And that's true, but `SamlInit` and `SamlFactory` were clear. It's when we decide to take the easy option and just have a `Utils` class that we lose clarity about what we each class is supposed to do.

However, as I said, personal preference. This definitely isn't a hill I'd die on. But those are the sorts of reasons why I tried to start with `SamlInit` and `SamlFactory` for this project.

**Comment by tvernum** [comment]
@elasticmachine update branch

File: x-pack/plugin/identity-provider/src/main/java/org/elasticsearch/xpack/idp/saml/sp/CloudServiceProvider.java
Comment (tvernum) [inline x-pack/plugin/identity-provider/src/main/java/org/elasticsearch/xpack/idp/saml/sp/CloudServiceProvider.java]: I still feel like this should be a URI. It _is_ a URI and I think we should consistently treat it like one.

File: x-pack/plugin/identity-provider/src/main/java/org/elasticsearch/xpack/idp/saml/sp/CloudServiceProvider.java
Comment (tvernum) [inline x-pack/plugin/identity-provider/src/main/java/org/elasticsearch/xpack/idp/saml/sp/CloudServiceProvider.java]: Can we make this a `Set` since we've done that with the index?

File: x-pack/plugin/identity-provider/src/main/java/org/elasticsearch/xpack/idp/saml/authn/SuccessfulAuthenticationResponseMessageBuilder.java
Comment (jkakavas) [reply x-pack/plugin/identity-provider/src/main/java/org/elasticsearch/xpack/idp/saml/authn/SuccessfulAuthenticationResponseMessageBuilder.java]: No need for anyone to die on this hill :) Given we've put in the time to climb it already though, I'll try and come up with something that you feel better with. I'll put the initialization logic in SamlInit to we can call `initialize()` on it and then move things back to a SamlFactory class which we can instantiate and call methods on.

File: x-pack/plugin/identity-provider/src/main/java/org/elasticsearch/xpack/idp/saml/sp/CloudServiceProvider.java
Comment (jkakavas) [reply x-pack/plugin/identity-provider/src/main/java/org/elasticsearch/xpack/idp/saml/sp/CloudServiceProvider.java]: I don't necessarily agree that this is important.

I reverted the change as you requested. There is value in validating that it is a URL (it's a URI but we should not be validating for a URI, as it can't be a URN ) and we will be doing this in the Registration API , no harm in doing it here too. Again, I think that for all intents and purposes when it comes to SAML, handling this as a URL in code - after validation - is unnecessary and it can potentially  introduce subtle bugs. 
For instance we need to take care so that when comparing `ACS` values , we are comparing string representations of the URLs and i.e. _not_ [`URL.equals`](https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/net/URL.html#equals(java.lang.Object)) which does something different. 

I didn't change the IDP configuration settings to be `Setting<URL>` too because of [the arguments here](https://github.com/elastic/elasticsearch/pull/51682#discussion_r373320101), the fact that we already validate them and finally that we have no good existing way to parse and handle `Setting<URL>` with optional values in a way that makes sense to me ( I can further elaborate if needed )

Comment (jkakavas) [reply x-pack/plugin/identity-provider/src/main/java/org/elasticsearch/xpack/idp/saml/sp/CloudServiceProvider.java]: I'm moving these to Setting<URL> in the IDP metadata PR

**Review by tvernum — APPROVED** [review]

**Pull request merged**
